File: 0000-0999/0131.PalindromePartitioning.go
func partition(s string) [][]string {
    ans := [][]string{}
    dfs(s, 0, []string{}, &ans)
    return ans
}

func dfs(s string, i int, path []string, ans *[][]string) {
    if i >= len(s) {
        tmp := make([]string, len(path))
        copy(tmp, path)
        *ans = append(*ans, tmp)
    }
    for j := i+1; j <= len(s); j++ {
        if isPalindrome(s[i:j]) {
            dfs(s, j, append(path, s[i:j]), ans)
        }
    }
}

func isPalindrome(s string) bool {
    for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
        if s[i] != s[j] {
            return false
        }
    }
    return true
}

File: 0000-0999/0131.PalindromePartitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        ans = []
        self.dfs(s, [], ans)
        return ans

    def dfs(self, s, path, ans):
        if not s:
            ans.append(path)
            return
        for i in xrange(1, len(s)+1):
            if self.isPal(s[:i]):
                self.dfs(s[i:], path+[s[:i]], ans)

    def isPal(self, s):
        return s == s[::-1]